api/src/wfl/service/datarepo.clj

```diff
@@ -117,3 +118,126 @@
+      poll-job))
+
+(defn list-snapshots
+  "List all Snapshots the caller has access to.
+   Hard-coded to return 999 pages for now.
+
+   Parameters
+   ----------
+   dataset-ids      - Optionally filter the result snapshots
+                      where provided datasets are source datasets.
+
+   Example
+   -------
+     (list-snapshots)
+     (list-snapshots \"48a51f71-6bab-483d-a270-3f9ebfb241cd\" \"85efdfea-52fb-4698-bee6-eef76104a7f4\")"
+  [& dataset-ids]
+  (letfn [(maybe [m k v] (if (seq v) (assoc m k {:datasetIds v}) m))]
+    (-> (http/get (repository "snapshots")
+                  (-> {:headers (auth/get-service-account-header)
+                       :query-params {:limit 999}}
+                      (maybe :query-params dataset-ids)))
+        util/response-body-json)))
+
+(defn delete-snapshot
+  "Delete the Snapshot with `snapshot-id`."
+  [dataset-id]
+  (-> (repository "snapshots/" dataset-id)
+      (http/delete {:headers (auth/get-service-account-header)})
+      util/response-body-json
+      :id
+      poll-job))
+
+(defn snapshot
+  "Query the DataRepo for the Snapshot with `snapshot-id`."
+  [snapshot-id]
+  (-> (repository "snapshots/" snapshot-id)
+      (http/get {:headers (auth/get-service-account-header)})
+      util/response-body-json))
+
+;; visible for testing
+(defn compose-snapshot-query
+  "Helper function for composing row-id query payload from `dataset` and `table`,
+   given a date range specified by exclusive `start` and inclusive `end`.
+
+   Note if there are no matching rows between (start, end], TDR will throw
+   a 400 exception.
+
+   Note TDR prefixes datasets in BigQuery with `datarepo_`.
+
+   Parameters
+   ----------
+   _dataset   - Dataset information response from TDR.
+   table      - Name of the table in the dataset schema to query from.
+   start      - The start date object in the timeframe to query exclusively.
+   end        - The end date object in the timeframe to query inclusively."
+  [{:keys [name dataProject] :as _dataset} table start end]
+  (let [dataset-name (str "datarepo_" name)]
+    (->> [dataProject dataset-name table start end]
+         (apply format "SELECT
+                        datarepo_row_id
+                     FROM
+                        `%s.%s.%s`
+                     WHERE
+                        datarepo_ingest_date > '%s' AND datarepo_ingest_date <= '%s'"))))
+
+(defn ^:private legalize-tdr-columns
+  "Legalize TDR columns by stripping out columns that are Array[File] types, as
+   TDR does not support them yet."
+  [cols]
+  (-> cols
+      (->> (filter #(not (and (= (:datatype %) "fileref")
+                              (= (:array_of %) true)))))))
+
+(defn ^:private all-columns
+  "Parse out and legalize all column names for a `table` in TDR `dataset`."
+  [dataset table]
+  (-> dataset
+      (get-in [:schema :tables])
+      (->> (filter #(= (:name %) table)))
+      first
+      :columns
+      legalize-tdr-columns
+      (#(map :name %))))
```

Comment:
Hmm, good question, I struggled when adding this. I ended up having this simply because `all-columns` returns only column/field names (not `:datatype` or `:array_of`)that could be directly consumed by other functions. Putting the filter elsewhere will require this function to return more information than a vector of name strings, and the filter can be simply removed. (I was also expecting the bug to be fixed pretty soon. ) What's your thought? I'm OK with either way.